### PR TITLE
Associate synthesized pattern matching switch case variables with switch statement

### PR DESCRIPTION
**Pull request by tmat**
**Customer scenario**

Enables EnC of switch statements with pattern matching.

Previously the emitted synthesized variables were associated with the case labels. In order to allow reordering during EnC we need to associate them with the switch statement itself and use their unique type as a key when mapping local slots of the same kind and ordinal. 

Manually tested E2E editing of a method with switch statement and works well with various kinds edits (reordering cases, updating when clauses, etc.).

**Bugs this fixes:**

Fixes https://github.com/dotnet/roslyn/issues/17991

**Workarounds, if any**

None

**Risk**

Small. 

**Performance impact**

None.

**Is this a regression from a previous update?**

**Root cause analysis:**

**How was the bug found?**

Planned feature.

### CODE REVIEW

**Comment by ivanbasov** [comment]
:shipit:

**Comment by tmat** [comment]
test windows_debug_vs-integration_prtest  please

**Comment by tmat** [comment]
test windows_release_vs-integration_prtest  please

**Comment by tmat** [comment]
test windows_release_vs-integration_prtest  please

**Comment by tmat** [comment]
@gafter @jcouv Ping.

**Comment by tmat** [comment]
@MeiChin-Tsai For ask mode approval.

File: src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs [hunk 1]
```diff
@@ -1535,7 +1535,11 @@ private string GetLocalDebugName(ILocalSymbolInternal local, out LocalDebugId lo
                 var syntax = local.GetDeclaratorSyntax();
                 int syntaxOffset = _method.CalculateLocalSyntaxOffset(syntax.SpanStart, syntax.SyntaxTree);
 
-                int ordinal = _synthesizedLocalOrdinals.AssignLocalOrdinal(localKind, syntaxOffset);
+                // Synthesized locals emitted for switch case patterns are all associated with the switch statement 
+                // and have distinct types. We use theier types to match them, not the ordinal as the ordinal might
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStatement.cs]: Nit: s/theier/their/

**Comment by jcouv** [comment]
:shipit:

**Comment by tmat** [comment]
Test windows_debug_vs-integration_prtest please

**Pull request merged**